File: legacy_backend/aggregators/heise.py
"""Aggregator for Heise RSS feeds."""

import contextlib
import html as html_module
import json
import re
from urllib.parse import parse_qs, urlencode, urlparse, urlunparse

from bs4 import BeautifulSoup
from pydantic import BaseModel, Field

from .base import BaseAggregator, RawArticle, fetch_article_content
from .base.exceptions import ContentFetchError


class HeiseAggregatorConfig(BaseModel):
    id: str
    type: str = Field(pattern="^(managed|custom|social)$")
    name: str = Field(min_length=1)
    url: str = ""
    description: str = Field(min_length=1)
    wait_for_selector: str | None = None
    selectors_to_remove: list[str] = Field(default_factory=list)
    options: dict = Field(default_factory=dict)


class HeiseAggregator(BaseAggregator):
    """Aggregator for Heise.de (German tech news)."""

    id = "heise"
    type = "managed"
    name = "Heise"
    url = "https://www.heise.de/rss/heise.rdf"
    description = "Specialized aggregator for Heise.de (German tech news). Extracts article content, removes ads and tracking elements, and filters out premium content and image galleries."
    options = {
        "traverse_multipage": {
            "type": "boolean",
            "label": "Traverse multi-page articles",
            "help_text": "Fetch and inline all pages of multi-page articles into a single article",
            "default": False,
        },
        "max_comments": {
            "type": "integer",
            "label": "Maximum comments to extract",
            "help_text": "Number of comments to extract and inline at the end of articles (0 to disable)",
            "default": 0,
            "min": 0,
            "max": 100,
        },
    }

    wait_for_selector = "#meldung, .StoryContent"
    selectors_to_remove = [
        ".ad-label",
        ".ad",
        ".article-sidebar",
        "section",
        "a[name='meldung.ho.bottom.zurstartseite']",
        "a-img",
        ".a-article-header__lead",
        ".a-article-header__title",
        ".a-article-header__publish-info",
        ".a-article-header__service",
        "div[data-component='RecommendationBox']",
        ".opt-in__content-container",
        ".a-box",
        "iframe",
        ".a-u-inline",
        ".redakteurskuerzel",
        ".branding",
        "a-gift",
        "aside",
        "script",
        "style",
        "noscript",
        "footer",
        ".rte__list",
        "#wtma_teaser_ho_vertrieb_inline_branding",
    ]

    def __init__(self):
        super().__init__()
        HeiseAggregatorConfig(
            id=self.id,
            type=self.type,
            name=self.name,
            url=self.url,
            description=self.description,
            wait_for_selector=self.wait_for_selector,
            selectors_to_remove=self.selectors_to_remove,
            options=self.options,
        )

    def should_skip_article(self, article: RawArticle) -> tuple[bool, str | None]:
        """Skip unwanted content types."""
        skip_terms = [
            "die Bilder der Woche",
            "Produktwerker",
            "heise-Angebot",
            "#TGIQF",
            "heise+",
            "#heiseshow:",
            "Mein Scrum ist kaputt",
            "software-architektur.tv",
            "Developer Snapshots",
        ]

        if any(term in article.title for term in skip_terms):
            return True, f"Skipping filtered content: {article.title}"

        return super().should_skip_article(article)

    def fetch_article_html(self, article: RawArticle) -> str:
        """Fetch HTML, optionally using multi-page URL."""
        url = article.url

        # Convert to all-pages URL if option enabled
        if self.get_option("traverse_multipage", False):
            url = self._convert_to_all_pages_url(url)
            self.logger.info(f"Using all-pages URL: {url}")

        return fetch_article_content(
            url,
            use_cache=not self.force_refresh,
            wait_for_selector=self.wait_for_selector,
            timeout=self.fetch_timeout,
        )

    def _convert_to_all_pages_url(self, url: str) -> str:
        """Convert URL to 'all pages on one page' version."""
        parsed = urlparse(url)
        params = parse_qs(parsed.query)
        params["seite"] = ["all"]
        new_query = urlencode(params, doseq=True)
        return urlunparse(
            (
                parsed.scheme,
                parsed.netloc,
                parsed.path,
                parsed.params,
                new_query,
                parsed.fragment,
            )
        )

    def extract_content(self, article: RawArticle) -> None:
        """Extract content from #meldung or .StoryContent."""
        soup = BeautifulSoup(article.html, "html.parser")

        story = soup.select_one("#meldung, .StoryContent")
        if not story:
            self.logger.warning("Could not find #meldung or .StoryContent")
            return  # Keep full HTML

        content = BeautifulSoup(str(story), "html.parser")

        # Remove empty elements
        for tag in content.find_all(["p", "div", "span"]):
            if not tag.get_text(strip=True) and not tag.find("img"):
                tag.decompose()

        article.html = str(content)

    def process_article(self, article: RawArticle, is_first: bool = False) -> str:
        """Process article with optional comments extraction."""
        # Standard processing
        content = super().process_article(article, is_first)

        # Add comments if enabled
        max_comments = self.get_option("max_comments", 0)
        if max_comments > 0:
            self.logger.info(f"Extracting up to {max_comments} comments")
            try:
                # Get raw HTML again for comment extraction
                raw_html = self.fetch_article_html(article)
                comments_html, forum_url = self._extract_comments(
                    article.url, raw_html, max_comments
                )
                if comments_html:
                    content = f"{content}\n\n{comments_html}"
                elif forum_url:
                    # Show "no comments" message when enabled but none found
                    no_comments_html = f'<h3><a href="{forum_url}">Comments</a></h3>\n<p><em>No comments available for this article.</em></p>'
                    content = f"{content}\n\n{no_comments_html}"
            except ContentFetchError as e:
                self.logger.warning(
                    f"Failed to fetch article HTML for comment extraction: {e}"
                )
                # Comments are optional, continue without them
            except Exception as e:
                self.logger.error(f"Failed to extract comments: {e}")

        return content

    def _extract_comments(
        self, article_url: str, article_html: str, max_comments: int
    ) -> tuple[str, str | None]:
        """Extract comments from a Heise article.

        Returns:
            Tuple of (comments_html, forum_url)
        """
        # Extract forum URL from article HTML
        forum_url = self._extract_forum_url(article_html, article_url)
        if not forum_url:
            self.logger.info("No forum URL found in article")
            return "", None

        self.logger.info(f"Fetching comments from forum: {forum_url}")

        # Fetch the forum page
        try:
            html = fetch_article_content(
                forum_url,
                use_cache=not self.force_refresh,
                wait_for_selector="body",
                timeout=30000,
            )
        except ContentFetchError as e:
            self.logger.warning(f"Failed to fetch comments from {forum_url}: {e}")
            return "", forum_url
        except Exception as e:
            self.logger.warning(
                f"Unexpected error fetching comments from {forum_url}: {e}"
            )
            return "", forum_url

        soup = BeautifulSoup(html, "html.parser")

        # Try different comment selectors
        comment_selectors = [
            "li.posting_element",
            '[id^="posting_"]',
            ".posting",
            ".a-comment",
        ]

        comment_elements = []
        for selector in comment_selectors:
            elements = soup.select(selector)
            if elements:
                self.logger.info(
                    f"Found {len(elements)} comments using selector: {selector}"
                )
                comment_elements = elements
                break

        if not comment_elements:
            self.logger.info("No comments found in forum HTML")
            return "", forum_url

        # Extract and format comments
        comment_html_parts = [f'<h3><a href="{forum_url}">Comments</a></h3>']
        extracted_count = 0

        for i, element in enumerate(comment_elements[:max_comments]):
            try:
                author = "Unknown"
                is_list_item = element.name == "li"

                if is_list_item:
                    # Extract from list view
                    author_elem = element.select_one(
                        ".tree_thread_list--written_by_user, .pseudonym"
                    )
                    if author_elem:
                        author = author_elem.get_text(strip=True)

                    title_link = element.select_one("a.posting_subject")
                    if not title_link:
                        continue

                    title = title_link.get_text(strip=True)
                    content = f"<p>{html_module.escape(title)}</p>"
                    comment_url = title_link.get("href", "")
                else:
                    # Extract from full posting view
                    author_selectors = [
                        'a[href*="/forum/heise-online/Meinungen"]',
                        ".pseudonym",
                        ".username",
                        "strong",
                    ]
                    for selector in author_selectors:
                        author_elem = element.select_one(selector)
                        if author_elem:
                            author_text = author_elem.get_text(strip=True)
                            if author_text and len(author_text) < 50:
                                author = author_text
                                break

                    # Extract content
                    content = ""
                    content_selectors = [
                        ".text",
                        ".posting-content",
                        ".comment-body",
                        "p",
                    ]
                    for selector in content_selectors:
                        content_elem = element.select_one(selector)
                        if content_elem:
                            content = str(content_elem)
                            break

                    comment_id = element.get("id") or f"comment-{i}"
                    comment_url = f"{article_url}#{comment_id}"

                if not content or not content.strip():
                    continue

                # Format comment
                comment_html = f"""<blockquote>
<p><strong>{html_module.escape(author)}</strong> | <a href="{comment_url}">source</a></p>
<div>{content}</div>
</blockquote>
"""
                comment_html_parts.append(comment_html)
                extracted_count += 1

            except Exception as e:
                self.logger.warning(f"Error extracting comment {i}: {e}")
                continue

        if extracted_count == 0:
            return "", forum_url

        self.logger.info(
            f"Successfully extracted {extracted_count} comments from article"
        )
        return "\n".join(comment_html_parts), forum_url

    def _extract_forum_url(self, article_html: str, article_url: str) -> str | None:
        """Extract the forum URL from article HTML."""
        soup = BeautifulSoup(article_html, "html.parser")

        # Look for JSON-LD script tag
        for script in soup.find_all("script", type="application/ld+json"):
            with contextlib.suppress(json.JSONDecodeError, KeyError):
                data = json.loads(script.string)
                # Handle both single object and array
                if isinstance(data, list):
                    for item in data:
                        if "discussionUrl" in item:
                            discussion_url = item["discussionUrl"]
                            if discussion_url.startswith("/"):
                                parsed = urlparse(article_url)
                                return (
                                    f"{parsed.scheme}://{parsed.netloc}{discussion_url}"
                                )
                            return discussion_url
                elif "discussionUrl" in data:
                    discussion_url = data["discussionUrl"]
                    if discussion_url.startswith("/"):
                        parsed = urlparse(article_url)
                        return f"{parsed.scheme}://{parsed.netloc}{discussion_url}"
                    return discussion_url

        # Fallback: look for comment links in HTML
        comment_link = soup.find("a", href=re.compile(r"/forum/.*comment"))
        if comment_link and comment_link.get("href"):
            href = comment_link["href"]
            if href.startswith("/"):
                parsed = urlparse(article_url)
                return f"{parsed.scheme}://{parsed.netloc}{href}"
            return href

        return None


def aggregate(feed, force_refresh=False, options=None):
    aggregator = HeiseAggregator()
    return aggregator.aggregate(feed, force_refresh, options or {})

File: src/server/aggregators/heise.ts
/**
 * Heise aggregator.
 *
 * Specialized aggregator for Heise.de (German tech news).
 */

import { FullWebsiteAggregator } from './full_website';
import type { RawArticle } from './base/types';
import { fetchArticleContent } from './base/fetch';
import { extractContent } from './base/extract';
import { standardizeContentFormat } from './base/process';
import { sanitizeHtml } from './base/utils';
import { logger } from '../utils/logger';
import * as cheerio from 'cheerio';
import { ContentFetchError } from './base/exceptions';

export class HeiseAggregator extends FullWebsiteAggregator {
  override readonly id = 'heise';
  override readonly type: 'managed' | 'custom' | 'social' = 'managed';
  override readonly name = 'Heise';
  override readonly url = 'https://www.heise.de/rss/heise.rdf';
  override readonly description =
    'Heise.de - German technology news website covering IT, science, and digital culture.';

  override readonly waitForSelector = '#meldung, .StoryContent';
  override readonly selectorsToRemove = [
    '.ad-label',
    '.ad',
    '.article-sidebar',
    'section',
    "a[name='meldung.ho.bottom.zurstartseite']",
    'a-img',
    '.a-article-header__lead',
    '.a-article-header__title',
    '.a-article-header__publish-info',
    '.a-article-header__service',
    "div[data-component='RecommendationBox']",
    '.opt-in__content-container',
    '.a-box',
    'iframe',
    '.a-u-inline',
    '.redakteurskuerzel',
    '.branding',
    'a-gift',
    'aside',
    'script',
    'style',
    'noscript',
    'footer',
    '.rte__list',
    '#wtma_teaser_ho_vertrieb_inline_branding',
  ];

  override readonly options = {
    // Inherit options from FullWebsiteAggregator
    exclude_selectors: {
      type: 'string' as const,
      label: 'CSS selectors to exclude (one per line)',
      helpText:
        'Additional CSS selectors for elements to remove from content. Enter one selector per line.\n\nExample:\n.advertisement\n.social-share\nfooter\nscript',
      default: '',
      required: false,
      widget: 'textarea' as const,
    },
    ignore_title_contains: {
      type: 'string' as const,
      label: 'Ignore articles if title contains (one per line)',
      helpText:
        'Skip articles if the title contains any of these strings (case-insensitive). Enter one string per line.\n\nExample:\n[SPONSORED]\nAdvertisement\nPremium',
      default: '',
      required: false,
      widget: 'textarea' as const,
    },
    ignore_content_contains: {
      type: 'string' as const,
      label: 'Ignore articles if content contains (one per line)',
      helpText:
        'Skip articles if the title or content contains any of these strings (case-insensitive). Enter one string per line.\n\nExample:\npaywall\nsubscription required\nmembers only',
      default: '',
      required: false,
      widget: 'textarea' as const,
    },
    regex_replacements: {
      type: 'string' as const,
      label: 'Regex replacements (one per line)',
      helpText:
        'Apply regex replacements to article content. One replacement per line in format: pattern|replacement\n\nApplied sequentially after all other processing.\n\nExample:\nfoo|bar\n\\d{4}|YEAR\n^\\s+|  (remove leading spaces)\n\nNote: Use | to separate pattern from replacement. To include a literal |, escape it as \\|',
      default: '',
      required: false,
      widget: 'textarea' as const,
    },
    // Heise-specific options
    traverse_multipage: {
      type: 'boolean' as const,
      label: 'Traverse multi-page articles',
      helpText: 'Fetch and inline all pages of multi-page articles into a single article',
      default: false,
      required: false,
    },
    max_comments: {
      type: 'integer' as const,
      label: 'Maximum comments to extract',
      helpText: 'Number of comments to extract and inline at the end of articles (0 to disable)',
      default: 0,
      required: false,
      min: 0,
      max: 100,
    },
  };

  protected override shouldSkipArticle(article: RawArticle): boolean {
    const skipTerms = [
      'die Bilder der Woche',
      'Produktwerker',
      'heise-Angebot',
      '#TGIQF',
      'heise+',
      '#heiseshow:',
      'Mein Scrum ist kaputt',
      'software-architektur.tv',
      'Developer Snapshots',
    ];

    if (skipTerms.some(term => article.title.includes(term))) {
      logger.info({ title: article.title }, 'Skipping filtered content');
      return true;
    }

    return super.shouldSkipArticle(article);
  }

  override async aggregate(articleLimit?: number): Promise<RawArticle[]> {
    if (!this.feed) {
      throw new Error('Feed not initialized');
    }

    // Get options
    const traverseMultipage = this.getOption('traverse_multipage', false) as boolean;
    const maxComments = this.getOption('max_comments', 0) as number;

    // Call parent aggregate to get base articles
    const articles = await super.aggregate(articleLimit);

    // Process each article with Heise-specific logic
    for (const article of articles) {
      try {
        // Skip if article already exists (unless force refresh)
        if (this.isExistingUrl(article.url)) {
          logger.debug(
            {
              url: article.url,
              title: article.title,
              aggregator: this.id,
              step: 'skip_existing',
            },
            'Skipping existing article (will not fetch content)'
          );
          continue;
        }

        // Convert to all-pages URL if option enabled
        let articleUrl = article.url;
        if (traverseMultipage) {
          try {
            const url = new URL(articleUrl);
            url.searchParams.set('seite', 'all');
            articleUrl = url.toString();
            logger.info({ url: articleUrl }, 'Using all-pages URL');
          } catch (error) {
            logger.warn({ error, url: article.url }, 'Failed to convert to all-pages URL');
          }
        }

        // Fetch article HTML
        const html = await fetchArticleContent(articleUrl, {
          timeout: this.fetchTimeout,
          waitForSelector: this.waitForSelector,
        });

        // Extract content
        const extracted = extractContent(html, {
          selectorsToRemove: this.selectorsToRemove,
          contentSelector: '#meldung, .StoryContent',
        });

        // Process content
        const $ = cheerio.load(extracted);

        // Remove empty elements
        $('p, div, span').each((_, el) => {
          const $el = $(el);
          if (!$el.text().trim() && !$el.find('img').length) {
            $el.remove();
          }
        });

        let content = $.html();

        // Sanitize HTML (remove scripts, rename attributes)
        content = sanitizeHtml(content);

        // Add comments if enabled
        if (maxComments > 0) {
          try {
            logger.info({ url: articleUrl, maxComments }, 'Extracting comments');
            const commentsHtml = await this.extractComments(articleUrl, html, maxComments);
            if (commentsHtml) {
              content = `${content}\n\n${commentsHtml}`;
            }
          } catch (error) {
            logger.warn({ error, url: articleUrl }, 'Failed to extract comments');
            // Comments are optional, continue without them
          }
        }

        // Standardize format (add header image, source link)
        const generateTitleImage = this.feed?.generateTitleImage ?? true;
        const addSourceFooter = this.feed?.addSourceFooter ?? true;
        article.content = await standardizeContentFormat(
          content,
          article,
          article.url,
          generateTitleImage,
          addSourceFooter
        );
      } catch (error) {
        logger.error({ error, url: article.url }, 'Error processing Heise article');
        // Continue with original content if processing fails
      }
    }

    return articles;
  }

  /**
   * Extract comments from a Heise article.
   */
  private async extractComments(
    articleUrl: string,
    articleHtml: string,
    maxComments: number
  ): Promise<string | null> {
    // Extract forum URL from article HTML
    const forumUrl = this.extractForumUrl(articleHtml, articleUrl);
    if (!forumUrl) {
      logger.info({ url: articleUrl }, 'No forum URL found in article');
      return null;
    }

    logger.info({ forumUrl }, 'Fetching comments from forum');

    try {
      // Fetch the forum page
      const html = await fetchArticleContent(forumUrl, {
        timeout: 30000,
        waitForSelector: 'body',
      });

      const $ = cheerio.load(html);

      // Try different comment selectors
      const commentSelectors = ['li.posting_element', '[id^="posting_"]', '.posting', '.a-comment'];

      let commentElements: cheerio.Cheerio<any> | null = null;
      for (const selector of commentSelectors) {
        const elements = $(selector);
        if (elements.length > 0) {
          logger.info({ selector, count: elements.length }, 'Found comments using selector');
          commentElements = elements;
          break;
        }
      }

      if (!commentElements || commentElements.length === 0) {
        logger.info({ forumUrl }, 'No comments found in forum HTML');
        return null;
      }

      // Extract and format comments
      const commentHtmlParts: string[] = [`<h3><a href="${forumUrl}">Comments</a></h3>`];
      let extractedCount = 0;

      commentElements.slice(0, maxComments).each((i, element) => {
        try {
          const $el = $(element);
          let author = 'Unknown';
          const isListItem = element.tagName === 'li';

          if (isListItem) {
            // Extract from list view
            const authorElem = $el.find('.tree_thread_list--written_by_user, .pseudonym').first();
            if (authorElem.length) {
              author = authorElem.text().trim();
            }

            const titleLink = $el.find('a.posting_subject').first();
            if (!titleLink.length) {
              return;
            }

            const title = titleLink.text().trim();
            const content = `<p>${cheerio.load(title).text()}</p>`;
            const commentUrl = titleLink.attr('href') || '';

            commentHtmlParts.push(
              `<blockquote><p><strong>${author}</strong> | <a href="${commentUrl}">source</a></p><div>${content}</div></blockquote>`
            );
            extractedCount++;
          } else {
            // Extract from full posting view
            const authorSelectors = [
              'a[href*="/forum/heise-online/Meinungen"]',
              '.pseudonym',
              '.username',
              'strong',
            ];

            for (const selector of authorSelectors) {
              const authorElem = $el.find(selector).first();
              if (authorElem.length) {
                const authorText = authorElem.text().trim();
                if (authorText && authorText.length < 50) {
                  author = authorText;
                  break;
                }
              }
            }

            // Extract content
            const contentSelectors = ['.text', '.posting-content', '.comment-body', 'p'];
            let content = '';
            for (const selector of contentSelectors) {
              const contentElem = $el.find(selector).first();
              if (contentElem.length) {
                content = contentElem.html() || '';
                break;
              }
            }

            const commentId = $el.attr('id') || `comment-${i}`;
            const commentUrl = `${articleUrl}#${commentId}`;

            if (!content || !content.trim()) {
              return;
            }

            commentHtmlParts.push(
              `<blockquote><p><strong>${author}</strong> | <a href="${commentUrl}">source</a></p><div>${content}</div></blockquote>`
            );
            extractedCount++;
          }
        } catch (error) {
          logger.warn({ error, index: i }, 'Error extracting comment');
        }
      });

      if (extractedCount === 0) {
        return null;
      }

      logger.info({ extractedCount }, 'Successfully extracted comments');
      return commentHtmlParts.join('\n');
    } catch (error) {
      if (error instanceof ContentFetchError) {
        logger.warn({ error, forumUrl }, 'Failed to fetch comments');
      } else {
        logger.warn({ error, forumUrl }, 'Unexpected error fetching comments');
      }
      return null;
    }
  }

  /**
   * Extract the forum URL from article HTML.
   */
  private extractForumUrl(articleHtml: string, articleUrl: string): string | null {
    const $ = cheerio.load(articleHtml);

    // Look for JSON-LD script tag
    $('script[type="application/ld+json"]').each((_, el) => {
      try {
        const text = $(el).text();
        const data = JSON.parse(text);
        const items = Array.isArray(data) ? data : [data];

        for (const item of items) {
          if (item.discussionUrl) {
            let discussionUrl = item.discussionUrl;
            if (discussionUrl.startsWith('/')) {
              const url = new URL(articleUrl);
              discussionUrl = `${url.origin}${discussionUrl}`;
            }
            return discussionUrl;
          }
        }
      } catch {
        // Invalid JSON, skip
      }
    });

    // Fallback: look for comment links in HTML
    const commentLink = $('a[href*="/forum/"][href*="comment"]').first();
    if (commentLink.length) {
      let href = commentLink.attr('href') || '';
      if (href.startsWith('/')) {
        const url = new URL(articleUrl);
        href = `${url.origin}${href}`;
      }
      return href;
    }

    return null;
  }
}
